File: Core/geometry/bounding_box.ts
import { Vector } from "./vector";

export class BoundingBox {
    readonly is_empty: boolean;
    readonly width: number;
    readonly height: number;
    readonly top_left: Vector;
    readonly top_right: Vector;
    readonly bottom_left: Vector;
    readonly bottom_right: Vector;

    constructor(
        readonly min_x: number,
        readonly min_y: number,
        readonly max_x: number,
        readonly max_y: number,
    ) {
        const w = this.max_x - this.min_x;
        this.width = Number.isFinite(w) ? w : 0;

        const h = this.max_y - this.min_y;
        this.height = Number.isFinite(h) ? h : 0;

        this.top_left = new Vector(this.min_x, this.min_y);
        this.top_right = new Vector(this.max_x, this.min_y);
        this.bottom_left = new Vector(this.min_x, this.max_y);
        this.bottom_right = new Vector(this.max_x, this.max_y);

        this.is_empty = min_x > max_x;
    }

    static empty() {
        return new BoundingBox(Infinity, Infinity, -Infinity, -Infinity);
    }

    static from_vectors(points: Vector[]) {
        if (points.length === 0) {
            return BoundingBox.empty();
        }

        let min_x = Infinity,
            min_y = Infinity,
            max_x = -Infinity,
            max_y = -Infinity;

        for (const p of points) {
            min_x = Math.min(min_x, p.x);
            min_y = Math.min(min_y, p.y);
            max_x = Math.max(max_x, p.x);
            max_y = Math.max(max_y, p.y);
        }

        return new BoundingBox(min_x, min_y, max_x, max_y);
    }

    intersects(other: BoundingBox) {
        return (
            this.min_x <= other.max_x &&
            this.max_x >= other.min_x &&
            this.min_y <= other.max_y &&
            this.max_y >= other.min_y
        );
    }

    merge(others: BoundingBox[]) {
        return BoundingBox.merge([this, ...others]);
    }

    get center() {
        return new Vector(
            (this.min_x + this.max_x) / 2,
            (this.min_y + this.max_y) / 2,
        );
    }

    static merge(boxes: BoundingBox[]) {
        return BoundingBox.from_vectors(
            boxes
                .filter((b) => !b.is_empty)
                .flatMap((b) => [
                    b.top_left,
                    b.top_right,
                    b.bottom_left,
                    b.bottom_right,
                ]),
        );
    }
}

File: Rust/crates/geometry/src/structs/bounding_box.rs
use crate::Vector;

#[derive(Clone, Copy)]
pub struct BoundingBox([f64; 4]);

#[allow(unused)]
impl BoundingBox {
    pub fn new(x: f64, y: f64, w: f64, h: f64) -> Self {
        debug_assert!(x.is_finite() && y.is_finite() && w.is_finite() && h.is_finite());
        debug_assert!(w >= 0.0 && h >= 0.0);
        BoundingBox([x, y, w, h])
    }

    pub fn from_x1_y1_x2_y2(x1: f64, y1: f64, x2: f64, y2: f64) -> Self {
        BoundingBox::new(x1, y1, x2 - x1, y2 - y1)
    }

    pub fn from_vectors<I>(iter: I) -> Option<Self>
    where
        I: IntoIterator<Item = Vector>,
    {
        let mut min_x = f64::INFINITY;
        let mut min_y = f64::INFINITY;
        let mut max_x = -f64::INFINITY;
        let mut max_y = -f64::INFINITY;

        for vec in iter.into_iter() {
            min_x = min_x.min(vec.x());
            min_y = min_y.min(vec.y());
            max_x = max_x.max(vec.x());
            max_y = max_y.max(vec.y());
        }

        if min_x.is_finite() {
            Some(BoundingBox::from_x1_y1_x2_y2(min_x, min_y, max_x, max_y))
        } else {
            None
        }
    }

    #[allow(unused)]
    pub fn x(&self) -> f64 {
        self.0[0]
    }

    #[allow(unused)]
    pub fn y(&self) -> f64 {
        self.0[1]
    }

    pub fn x1(&self) -> f64 {
        self.0[0]
    }

    pub fn y1(&self) -> f64 {
        self.0[1]
    }

    pub fn x2(&self) -> f64 {
        self.0[0] + self.0[2]
    }

    pub fn y2(&self) -> f64 {
        self.0[1] + self.0[3]
    }

    #[allow(unused)]
    pub fn width(&self) -> f64 {
        self.0[2]
    }

    #[allow(unused)]
    pub fn height(&self) -> f64 {
        self.0[3]
    }

    pub fn distance(self, other: BoundingBox) -> f64 {
        let dx = f64::max(
            0.0,
            f64::max(self.x1() - other.x2(), other.x1() - self.x2()),
        );
        let dy = f64::max(
            0.0,
            f64::max(self.y1() - other.y2(), other.y1() - self.y2()),
        );

        (dx * dx + dy * dy).sqrt()
    }

    pub fn contains(&self, v: Vector) -> bool {
        self.x1() <= v.x() && self.x2() >= v.x() && self.y1() <= v.y() && self.y2() >= v.y()
    }

    pub fn contains_properly(&self, v: Vector) -> bool {
        self.x1() < v.x() && self.x2() > v.x() && self.y1() < v.y() && self.y2() > v.y()
    }
}
